iothub_client/samples/pnp/pnp_simple_thermostat/pnp_simple_thermostat.c

```diff
@@ -0,0 +1,441 @@
+static const char g_ModelId[] = "dtmi:com:example:Thermostat;1";
+
+// JSON fields from desired property to retrieve.
+static const char PnP_JsonPropertyVersion[] = "$version";
+static const char PnP_JsonTargetTemperature[] = "targetTemperature";
+
+// Name of command this component supports to get report information
+static const char g_GetMinMaxReport[] = "getMaxMinReport";
+// Name of json field to parse for "since". (Note the "since" is assumed as only field of the value)
+static const char g_SinceJsonCommandSetting[] = "commandRequest.value";
+// Return codes for device methods and desired property responses
+static int g_statusSuccess = 200;
+static int g_statusBadFormat = 400;
+static int g_statusNotFoundStatus = 404;
+static int g_statusInternalError = 500;
+
+// The default temperature to use before any is set.
+#define DEFAULT_TEMPERATURE_VALUE 30
+// Current temperature of the thermostat.
+static double g_currentTemperature = DEFAULT_TEMPERATURE_VALUE;
+// Minimum temperature thermostat has been at during current execution run.
+static double g_minTemperature = DEFAULT_TEMPERATURE_VALUE;
+// Maximum temperature thermostat has been at during current execution run.
+static double g_maxTemperature = DEFAULT_TEMPERATURE_VALUE;
+// Number of times temperature has been updated, counting the initial setting as 1.  Used to determine average temperature.
+static int g_numTemperatureUpdates = 1;
+// Total of all temperature updates during current exceution run.  Used to determine average temperature.
+static double g_allTemperatures = DEFAULT_TEMPERATURE_VALUE;
+
+// snprintf format for building getMaxMinReport
+static const char g_minMaxCommandResponseFormat[] = "{ \"maxTemp\": %.2f, \"minTemp\": %.2f, \"avgTemp\": %.2f, \"startTime\": \"%s\", \"endTime\": \"%s\" }";
+
+// Format string for sending temperature telemetry
+static const char g_TemperatureTelemetryBodyFormat[] = "{ \"temperature\":  %.02f }";
+
+// Response description is an optional, human readable message including more information
+// about the setting of the temperature.  Because we accept all temperature requests, we 
+// always indicate a success.  An actual sensor could optionally return information about
+// a temperature being out of range or a mechanical issue on the device on error cases.
+static const char g_temperaturePropertyResponseDescription[] = "success";
+
+//
+// CopyTwinPayloadToString takes the twin payload data, which arrives as a potentially non-NULL terminated string, and creates
+// a new copy of the data with a NULL terminator.  The JSON parser this sample uses, parson, only operates over NULL terminated strings.
+//
+static char* CopyTwinPayloadToString(const unsigned char* payload, size_t size)
+{
+    char* jsonStr;
+
+    if ((jsonStr = (char*)malloc(size+1)) == NULL)
+    {
+        LogError("Unable to allocate %lu size buffer", (unsigned long)(size+1));
+    }
+    else
+    {
+        memcpy(jsonStr, payload, size);
+        jsonStr[size] = 0;
```

Comment:
nit: purely preferential but `\0` since it's a `char`. I always forget if we have a team consensus here

Reply:
We're inconsistent but the `\0` is "more correct" so I'm changing to be that.